Zend/zend_compile.c

```diff
@@ -3629,16 +3579,114 @@ void zend_compile_break_continue(zend_ast *ast) /* {{{ */
 }
 /* }}} */
 
+void zend_resolve_goto_label(zend_op_array *op_array, znode *label_node, zend_op *pass2_opline) /* {{{ */
+{
+	zend_label *dest;
+	int current, distance, free_vars;
+	zval *label;
+	znode *loop_var = NULL;
+
+	if (pass2_opline) {
+		label = RT_CONSTANT(op_array, pass2_opline->op2);
+	} else {
+		label = &label_node->u.constant;
+	}
+	if (CG(context).labels == NULL ||
+	    (dest = zend_hash_find_ptr(CG(context).labels, Z_STR_P(label))) == NULL) {
+
+		if (pass2_opline) {
+			CG(in_compilation) = 1;
+			CG(active_op_array) = op_array;
+			CG(zend_lineno) = pass2_opline->lineno;
+			zend_error_noreturn(E_COMPILE_ERROR, "'goto' to undefined label '%s'", Z_STRVAL_P(label));
+		} else {
+			/* Label is not defined. Delay to pass 2. */
+			zend_op *opline;
+
+			current = CG(context).current_brk_cont;
+			while (current != -1) {
+				if (CG(context).brk_cont_array[current].start >= 0) {
+					zend_emit_op(NULL, ZEND_NOP, NULL, label_node);
+				}
+				current = CG(context).brk_cont_array[current].parent;
+			}
+			opline = zend_emit_op(NULL, ZEND_GOTO, NULL, label_node);
+			opline->extended_value = CG(context).current_brk_cont;
+			return;
+		}
+	}
+
+	zval_dtor(label);
+	ZVAL_NULL(label);
+
+	/* Check that we are not moving into loop or switch */
+	if (pass2_opline) {
+		current = pass2_opline->extended_value;
+	} else {
+		current = CG(context).current_brk_cont;
+	}
+	if (!pass2_opline) {
+		loop_var = zend_stack_top(&CG(loop_var_stack));
+	}
+	for (distance = 0, free_vars = 0; current != dest->brk_cont; distance++) {
+		if (current == -1) {
+			if (pass2_opline) {
+				CG(in_compilation) = 1;
+				CG(active_op_array) = op_array;
+				CG(zend_lineno) = pass2_opline->lineno;
+			}
+			zend_error_noreturn(E_COMPILE_ERROR, "'goto' into loop or switch statement is disallowed");
```

Comment:
this is unrelated, and we won't change this now.

On Tue, Jul 7, 2015 at 6:43 AM, Bob Weinand <notifications@github.com>
wrote:

> In Zend/zend_compile.c
> <https://github.com/php/php-src/pull/1393#discussion_r34005934>:
>
> > +	if (pass2_opline) {
> > +		current = pass2_opline->extended_value;
> > +	} else {
> > +		current = CG(context).current_brk_cont;
> > +	}
> > +	if (!pass2_opline) {
> > +		loop_var = zend_stack_top(&CG(loop_var_stack));
> > +	}
> > +	for (distance = 0, free_vars = 0; current != dest->brk_cont; distance++) {
> > +		if (current == -1) {
> > +			if (pass2_opline) {
> > +				CG(in_compilation) = 1;
> > +				CG(active_op_array) = op_array;
> > +				CG(zend_lineno) = pass2_opline->lineno;
> > +			}
> > +			zend_error_noreturn(E_COMPILE_ERROR, "'goto' into loop or switch statement is disallowed");
>
> While we are here, we could allow goto into a switch statement; we'd just
> have to set the switch temporary to IS_NULL via some opcode first. (just
> not into loop...) [I've honestly missed it at times.]
>
> Just a suggestion, maybe it's also way off-scope to this PR...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1393/files#r34005934>.
>